extended/src/main/java/io/kubernetes/client/extended/controller/DefaultController.java

```diff
@@ -0,0 +1,192 @@
+package io.kubernetes.client.extended.controller;
+
+import io.kubernetes.client.extended.controller.reconciler.Reconciler;
+import io.kubernetes.client.extended.wait.Wait;
+import io.kubernetes.client.extended.workqueue.RateLimitingQueue;
+import java.time.Duration;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultController implements Controller {
+
+  private static final Logger log = LoggerFactory.getLogger(DefaultController.class);
+
+  private Reconciler reconciler;
+  private String name;
+  private int workerCount;
+  private ExecutorService workerThreadPool;
+
+  private final RateLimitingQueue<Reconciler.Request> workQueue;
+
+  private final Supplier<Boolean>[] readyFuncs;
+  private Duration readyTimeout = Duration.ofSeconds(30);
+  private Duration readyCheckInternal = Duration.ofSeconds(30);
+
+  public DefaultController(
+      Reconciler reconciler,
+      RateLimitingQueue<Reconciler.Request> workQueue,
+      Supplier<Boolean>... readyFuncs) {
+    this.reconciler = reconciler;
+    this.workQueue = workQueue;
+    this.readyFuncs = readyFuncs;
+  }
+
+  @Override
+  public void run() {
+    if (workerCount <= 0) {
+      log.error("Fail to start controller {}: worker count must be positive.");
+    }
+    if (workerThreadPool == null) {
+      log.error("Fail to start controller {}: missing worker thread-pool.");
+    }
+
+    boolean synced = true;
+    if (this.readyFuncs.length > 0) {
+      synced =
+          Wait.poll(
+              readyCheckInternal,
+              readyTimeout,
+              () -> {
+                boolean ready = true;
+                for (Supplier<Boolean> cacheHasSyncedFunc : this.readyFuncs) {
+                  ready = ready && cacheHasSyncedFunc.get();
+                }
+                return ready;
+              });
+    }
+
+    if (!synced) {
+      log.error("Timed out waiting for cache to be synced.");
+      return;
+    }
+
+    CountDownLatch latch = new CountDownLatch(workerCount);
+    for (int i = 0; i < this.workerCount; i++) {
+      final int workerIndex = i;
+      workerThreadPool.submit(
+          () -> {
+            if (log.isDebugEnabled()) {
+              log.debug("Starting controller {} worker {}..", this.name, workerIndex);
+            }
+            this.worker();
+            latch.countDown();
+            if (log.isDebugEnabled()) {
+              log.debug("Exiting controller {} worker {}..", this.name, workerIndex);
+            }
+          });
+    }
+    try {
+      if (log.isDebugEnabled()) {
+        log.debug("Controller {} bootstrapping..", this.name);
+      }
+      latch.await();
+    } catch (Throwable t) {
```

Comment:
Don't blindly catch `Throwable()` catch the specific exception(s) you care about.

Reply:
updated

Reply:
Looks like you're still catching `Throwable` ?

Reply:
As above, don't just catch `Throwable`

Reply:
updated, ptal